pkg/virt-controller/watch/vsock_test.go

```diff
@@ -0,0 +1,83 @@
+package watch
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/controller"
+	"kubevirt.io/kubevirt/tests"
+)
+
+var _ = Describe("VSOCK", func() {
+	var m *cidsMap
+
+	newRandomVMIsWithORWithoutVSOCK := func(totalVMINum, vsockVMINum int) []*virtv1.VirtualMachineInstance {
+		var vmis []*virtv1.VirtualMachineInstance
+		for i := 0; i < totalVMINum; i++ {
+			vmi := tests.NewRandomVMI()
+			if i < vsockVMINum {
+				cid := uint32(i + 3)
+				vmi.Status.VSOCKCID = &cid
+			}
+			vmis = append(vmis, vmi)
+		}
+		return vmis
+	}
+
+	BeforeEach(func() {
+		m = &cidsMap{
+			cids:    map[string]uint32{},
+			reverse: map[uint32]string{},
+		}
+	})
+
+	DescribeTable("Syncing CIDs from exsisting VMIs",
+		func(totalVMINum, vsockVMINum int) {
+			vmis := newRandomVMIsWithORWithoutVSOCK(totalVMINum, vsockVMINum)
+			for _, vmi := range vmis {
+				fmt.Println(vmi.Status.VSOCKCID)
+			}
+			m.Sync(vmis)
+			Expect(m.cids).To(HaveLen(vsockVMINum))
+			Expect(m.reverse).To(HaveLen(vsockVMINum))
+			for _, vmi := range vmis {
+				if vmi.Status.VSOCKCID == nil {
+					Expect(m.cids).NotTo(HaveKey(controller.VirtualMachineInstanceKey(vmi)))
+				} else {
+					Expect(m.cids).To(HaveKey(controller.VirtualMachineInstanceKey(vmi)))
+					Expect(m.reverse).To(HaveKey(*vmi.Status.VSOCKCID))
+				}
+			}
+		},
+		Entry("When all VMIs have VSOCK enabled", 3, 3),
+		Entry("When some VMIs have VSOCK enabled", 5, 3),
+	)
+
+	It("should allocate CIDs for VMIs", func() {
+		vmis := newRandomVMIsWithORWithoutVSOCK(5, 0)
+		for _, vmi := range vmis {
+			Expect(m.Allocate(vmi)).To(Succeed())
+			Expect(vmi.Status.VSOCKCID).NotTo(BeNil())
+			Expect(*vmi.Status.VSOCKCID).To(BeNumerically(">=", 3))
+		}
+		Expect(m.cids).To(HaveLen(len(vmis)))
+		Expect(m.reverse).To(HaveLen(len(vmis)))
+	})
+
+	It("should remove CIDs for VMIs", func() {
```

Comment:
I would love to see a negative test, where we get confirmation that we are not running in an infinite loop if nothing is left.

Reply:
Let's also acquire, I don't know 100 CIDs, and ensure that they are indeed all different.

Reply:
Done